Check-in scheduler

unflxw opened the pull request:
Part of https://github.com/appsignal/integration-guide/issues/169 (heartbeat check-in support in integrations) -- this does not implement heartbeat check-ins, but improves the existing cron check-ins implementation in a way that the heartbeat check-ins will benefit from.

---

Implement a check-in event scheduler, which schedules check-in events
to be transmitted in a separate thread, with a minimum wait period
of ten seconds between requests, and a wait period of a tenth of a
second before the first request, referred to as "debounce" periods.

This is a relatively minor improvement for the existing cron
check-ins, but it is a requirement for the heartbeat check-ins, both
to avoid slowing down customers' applications with blocking requests,
and to avoid misuse of the feature from spamming our servers.

The scheduler also acts as a deduplicator, removing "similar enough"
check-in events -- again, not particularly interesting for cron
check-ins, but a requirement to minimise damage when heartbeat
check-ins are misused.

Implement support for NDJSON payloads in the transmitter in order to
send all the check-ins in a single request.

The whole setup is lazily instantiated, so no threads will be spawned
until a check-in event actually needs to be transmitted. Internally,
a queue is used to communicate with the thread, which is suspended
until a message is pushed into the queue.

A second thread is used to keep track of the debounce period, and is
killed after a debounce period does not cause any events to be sent.

When `Appsignal.stop` is called, the scheduler is gracefully stopped,
blocking the call to `Appsignal.stop` until it has transmitted all
enqueued check-in events. During this time, new check-in events
cannot be scheduled.

matsimitsu commented on lib/appsignal/check_in/debouncer.rb:
```diff
@@ -0,0 +1,190 @@
+        begin
+          response = CheckIn.transmitter.transmit(events, :format => :ndjson)
+
+          if response.code.to_i >= 200 && response.code.to_i < 300
```
```suggestion
          if (200...300).include response.code.to_i
```

matsimitsu commented on lib/appsignal/check_in/debouncer.rb:
```diff
@@ -0,0 +1,190 @@
+
+      def run
+        loop do
+          events = @queue.pop
```
Won't this transmit a single event? (I think `queue.pop` only pops one. value)

unflxw replied:
You're correct that `queue.pop` only pops one value, but [the value it pops is an array of events](https://github.com/appsignal/appsignal-ruby/pull/1245/files/97e0cf6ac48b61e37ae89aba801c5b9d885714a6#diff-777a84febbe13fff07d37faf6813fbeb192a5011bdb4622d808ef5204e731426R180-R183) to be sent.

matsimitsu reviewed (changes requested):
Nice! Looks good :)

tombruijn commented on .changesets/send-check-ins-concurrently.md:
```diff
@@ -0,0 +1,8 @@
+
+Send check-ins concurrently. When calling `Appsignal::CheckIn.cron`, instead of blocking the current thread while the check-in events are sent, schedule them to be sent in a separate thread.
+
+When shutting down your application, make sure to call `Appsignal.stop` to block until all scheduled check-ins have been sent.
```
```suggestion
When shutting down your application manually, call `Appsignal.stop` to block until all scheduled check-ins have been sent.
```

tombruijn commented on lib/appsignal/cli/diagnose.rb:
```diff
@@ -150,7 +150,7 @@ def transmit_report_to_appsignal
             Appsignal.config
           )
-          response = transmitter.transmit(:diagnose => data)
+          response = transmitter.transmit({ :diagnose => data })
```
This change is not needed? It makes no difference.

tombruijn replied:
You're right! The mix of positional and keyword arguments threw me off.

tombruijn commented on lib/appsignal/transmitter.rb:
```diff
@@ -53,17 +54,39 @@ def uri
+      when :ndjson
+        NDJSON_CONTENT_TYPE
+      else
+        raise ArgumentError, "Unknown format: #{format}"
```
```suggestion
        raise ArgumentError, "Unknown format for Content-Type header: '#{format}'"
```

unflxw replied:
(flipped around the wording to "Unknown Content-Type header for format: #{format}")

tombruijn commented on lib/appsignal/transmitter.rb:
```diff
@@ -53,17 +54,39 @@ def uri
+      when :ndjson
+        Appsignal::Utils::NDJSON.generate(payload)
+      else
+        raise ArgumentError, "Unknown format: #{format}"
```
```suggestion
        raise ArgumentError, "Unknown format for request body: '#{format}'"
```

unflxw replied:
(flipped around the wording to "Unknown body generator for format: #{format}")

tombruijn commented on lib/appsignal/check_in/debouncer.rb:
```diff
@@ -0,0 +1,190 @@
+        @transmitted = 0
+      end
+
+      def send(event)
```
I would rename this method so you're not overwriting Ruby's `send` method. Like `transmit` and rename the private `transmit` method.

unflxw replied:
Done! Went with `schedule` instead.

tombruijn commented on lib/appsignal/check_in/debouncer.rb:
```diff
@@ -0,0 +1,190 @@
+          # If we're not already waiting to be awakened from a scheduled
+          # debounce, schedule a short debounce, which will push the events
+          # to the queue and schedule a long debounce.
+          schedule_waker(SHORT_DEBOUNCE_SECONDS) if @waker.nil?
```
```suggestion
          schedule_waker(INITIAL_DEBOUNCE_SECONDS) if @waker.nil?
```

Would it be more accurate to say it's the initial sleep/debounce time?

unflxw replied:
Good idea, makes a lot more sense semantically! We should also rename the other one (`LONG_DEBOUNCE_SECONDS`) in turn. How about `FOLLOWING_DEBOUNCE_SECONDS` or `AFTER_TRANSMIT_DEBOUNCE_SECONDS`?

tombruijn replied:
We use `ITERATION_IN_SECONDS` in the probes system. Could work here too like `DEBOUNCE_TIME_IN_SECONDS`

unflxw replied:
I went with `BETWEEN_TRANSMISSIONS_DEBOUNCE_SECONDS` 🤷

tombruijn commented on spec/lib/appsignal/check_in/cron_spec.rb:
okay but after this PR I'm refactoring these specs 😅 I had made a tiny start on this, but nothing I can't rebase.

tombruijn commented on spec/lib/appsignal/check_in/debouncer_spec.rb:
```diff
@@ -0,0 +1,483 @@
+      stub_const("Appsignal::CheckIn::Debouncer::LONG_DEBOUNCE_SECONDS", 10)
+
+      take_at_most(0.1) do
+        expect { subject.stop }.not_to raise_error
```
This assertion doesn't do much. If it errors without this wrapped assertion, it will bubble up and stop the test with the error as well.

```suggestion
        subject.stop
```

unflxw replied:
Yeah, `expect { ... }.not_to raise_error` never changes the outcome -- it's only there to indicate intent.

unflxw replied:
After taking a second look at the tests that use this `expect { ... }.not_to raise_error` expectation, I'm leaning towards leaving it there, for the above reason -- otherwise, these errors have no expectations, and it's not very clear what their purpose is. Leave a comment if you disagree.

tombruijn commented on spec/support/helpers/take_at_most_helper.rb:
```diff
@@ -0,0 +1,21 @@
+  #  # Assert that it takes at most 1 second to run the block
+  #  take_at_most(1) { sleep 0.5 }
+  #
+  # @param time [Integer] The maximum amount of time the block is allowed to
```
I see it being used with a float too.
```suggestion
  # @param time [Integer, Float] The maximum amount of time the block is allowed to
```

unflxw commented:
Updates since last review:

- Renamed "debouncer" to "scheduler", reducing the ugliness of the naming by 280%.
- Addressed feedback -- see comments.
